### DISCUSSION TITLE
How to use Response<BoxBody> with and without json payload in actix_http v3?

### DISCUSSION THREAD

**Discussion by pmalek-sumo** [actix/actix-web#2676] [Q&A] [answered]
It seems that with upgrade to actix_http v3 we can use e.g. `actix_http::Response::bad_request()` (which I understand is a new way of doing `actix_http::Response::BadRequest().body(Body::Empty)`) which returns `Response<BoxBody>` but then we cannot return json payload in the same handler as we used to with `actix_http::Response::Ok().json(__MY_STRUCT_IMPLEMENTING_derive_Deserialize__)`.

I've tried with

```rust
actix_http::Response::new(actix_http::StatusCode::OK).set_body(
    web::Json( __MY_STRUCT_IMPLEMENTING_derive_Deserialize__ {
        ...
    })
    ).map_into_boxed_body()
```

but then I get:

```
the trait bound `Json<CollectorRegisterRespone>: MessageBody` is not satisfied
the trait `MessageBody` is not implemented for `Json<CollectorRegisterRespone>`rustc[E0277](https://doc.rust-lang.org/error-index.html#E0277)
[response.rs(209, 12): ]()required by a bound in `actix_http::Response::<B>::map_into_boxed_body`
```

Relevant piece of code I'm trying to migrate:
https://github.com/SumoLogic/sumologic-kubernetes-tools/blob/3b0b24ea495726a6e7ad96fc3338d144c16a31c1/src/rust/receiver-mock/src/router/api.rs#L7-L43

**Comment by robjtede** [comment, marked as answer]
Firstly familiarize yourself with the body type changes in v4 [explained here](https://github.com/actix/actix-web/blob/master/actix-web/MIGRATION-4.0.md#response-body-types). Also it would be helpful to use `HttpResponse` from Actix Web instead of the `actix-http` types in your case since it contains fewer high-level things now.

Eg:
```diff
-           None => return actix_http::Response::BadRequest().body(Body::Empty),
+           None => return actix_web::HttpResponse::BadRequest().finish(),

-       actix_http::Response::Ok().json(CollectorRegisterRespone {
+       actix_web::HttpResponse::Ok().json(CollectorRegisterRespone {
```

**Comment by pmalek-sumo** [comment]
Thanks @robjtede! That worked and I wasn't aware of this guide 😬
